test/lit/passes/code-folding-eh.wast

```diff
@@ -0,0 +1,291 @@
+          (return (i32.const 0))
+        )
+        (br $tryend)
+      )
+      (call $foo)
+      (call $foo)
+      (call $foo)
+      (call $foo)
+      (return (i32.const 0))
+    )
+    (i32.const 0)
+  )
+
+  ;; CHECK:      (func $foo-i32 (type $0) (result i32)
+  ;; CHECK-NEXT:  (i32.const 0)
+  ;; CHECK-NEXT: )
+  (func $foo-i32 (result i32)
+    (i32.const 0)
+  )
+
+  ;; CHECK:      (func $try-call-optimize-terminating-tails-call-return (type $0) (result i32)
+  ;; CHECK-NEXT:  (block $tryend
+  ;; CHECK-NEXT:   (block $catch
+  ;; CHECK-NEXT:    (try_table (catch_all $catch)
+  ;; CHECK-NEXT:     (drop
+  ;; CHECK-NEXT:      (i32.const 1)
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:     (drop
+  ;; CHECK-NEXT:      (i32.const 1)
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:     (return
+  ;; CHECK-NEXT:      (call $foo-i32)
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (br $tryend)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (drop
+  ;; CHECK-NEXT:    (i32.const 1)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (drop
+  ;; CHECK-NEXT:    (i32.const 1)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (return
+  ;; CHECK-NEXT:    (call $foo-i32)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (i32.const 0)
+  ;; CHECK-NEXT: )
+  (func $try-call-optimize-terminating-tails-call-return (result i32)
+    (block $tryend
+      (block $catch
+        (try_table (catch_all $catch)
+          (drop (i32.const 1))
+          (drop (i32.const 1))
+          ;; Cannot be folded out of the try because it might throw.
+          (return (call $foo-i32))
+        )
+        (br $tryend)
+      )
+      (drop (i32.const 1))
+      (drop (i32.const 1))
+      (return (call $foo-i32))
+    )
+    (i32.const 0)
+  )
+
+  ;; CHECK:      (func $try-call-optimize-terminating-tails-return-call (type $0) (result i32)
+  ;; CHECK-NEXT:  (block $folding-inner0
+  ;; CHECK-NEXT:   (return
+  ;; CHECK-NEXT:    (block (result i32)
+  ;; CHECK-NEXT:     (block $tryend
+  ;; CHECK-NEXT:      (block $catch
+  ;; CHECK-NEXT:       (try_table (catch_all $catch)
+  ;; CHECK-NEXT:        (br $folding-inner0)
+  ;; CHECK-NEXT:       )
+  ;; CHECK-NEXT:       (br $tryend)
+  ;; CHECK-NEXT:      )
+  ;; CHECK-NEXT:      (br $folding-inner0)
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:     (i32.const 0)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (i32.const 1)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (i32.const 1)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (i32.const 1)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (return_call $foo-i32)
+  ;; CHECK-NEXT: )
+  (func $try-call-optimize-terminating-tails-return-call (result i32)
+    (block $tryend
+      (block $catch
+        (try_table (catch_all $catch)
+          (drop (i32.const 1))
+          (drop (i32.const 1))
+          (drop (i32.const 1))
+          (return_call $foo-i32)
```

Comment:
Can you add a line of comment on why this can be optimized, both here and the `eh-legacy` one, while you are at this? (`return_call` means calling the callee after we return the current function, so it will not be caught by `try`/`try_catch`, so we can optimize it)